### DISCUSSION TITLE
Typescript - Problems with setting up Scene and Stage

### DISCUSSION THREAD

**Discussion by ssj4maiko** [telegraf/telegraf#1455] [Q&A]
I already have a bot working, it was made with Telegram 3, and now I'm making another one, this time with Typescript and Telegram 4. Typescript meaning that I need to get the types right and all the stuff.

On the Telegram 3 Bot, the working code is something like this

```js
// ...
const initialContext = async (ctx) => {
	// Does stuff
};
const initialScene = new BaseScene('initial');
initialScene.enter(initialContext);
const secondContext = async (ctx) => {
	// Does validation that will redirect to second or return
};
const secondScene = new BaseScene('second');
secondScene.enter(secondContext);

const stage = new Stage;
stage.register(initialScene);
stage.register(secondScene);

bot.use(session);
bot.use(stage.middleware());
// ...
```

From my understanding, Scenes are small chunks of the Stage.

And this is my new code

```ts
import bot from '../bot';
import redisSession from '../../services/redis';
import { Context, Scenes } from 'telegraf';
import Update from 'typegram/update';

const firstScene = new Scenes.BaseScene('SceneInitial');

firstScene.enter((ctx) => {
	console.log('FIRST SCENE',ctx);
});

const StartupStage = new Scenes.Stage([firstScene]);
StartupStage.register(firstScene);

bot.use(redisSession);
bot.use(StartupStage.middleware())
export default StartupStage;
```
What I get, however, are different errors.
If I don't add any type to "ctx" in firstScene.enter, it gives no immediate error, but then the Scenes.Stage part will accuse:
`Type 'BaseScene<Context<Update>>' is not assignable to type 'BaseScene<SessionContext<SceneSession<SceneSessionData>> & { scene: SceneContextScene<Context<Update>, SceneSessionData>; }>'.`
If I add this type to the previous ctx, the I will get an error on the "firstScene.enter" part:
`Argument of type '(ctx: Scenes.BaseScene<SessionContext<Scenes.SceneSession<Scenes.SceneSessionData>> & {    scene: Scenes.SceneContextScene<Context<Update.Update>, Scenes.SceneSessionData>;}>) => void' is not assignable to parameter of type 'Middleware<Context<Update>>'.`
At the same time, `bot.use(StartupStage.middleware())` Also gives me type related errors, which may be fixed if I can fix the 2 previous ones.

So I ask, what am I doing wrong? Or what am I not doing that I should have?

**Comment by MichaelAroundTheWorld** [comment]
Use custom ContexType. Add following
```
import { Context, Scenes, session } from "telegraf";
import { Update } from "@telegraf/types";


interface MySession extends Scenes.SceneSessionData {

}


interface MyContext extends Context<Update> {
  session: MySession & Scenes.SceneSession<MySession>;
  scene: Scenes.SceneContextScene<MyContext, MySession>; 
}

export { MyContext, MySession };
```
Then use this type in your scene and stages
`const startScene = new Scenes.BaseScene<MyContext>("start");`

```
    const bot = new Telegraf<MyContext>(process.env.TELEGRAM_TOKEN);

    const stages = new Scenes.Stage<MyContext>([startScene]);
    bot.use(session());
    bot.use(stages.middleware());

    bot.start((ctx) => ctx.scene.enter("start"));
    bot.launch();
```
